parachain-upgrade/src/lib.rs

```diff
@@ -0,0 +1,588 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! Enable parachain validation function upgrades.
+//!
+//! Allow a user to determine when a parachain validation function upgrade
+//! is legal, and perform the upgrade, triggering runtime events
+//! for both storing and applying the new validation function.
+//!
+//! Depends on no external pallets or traits.
+//!
+//! This pallet depends on certain environmental conditions provided by
+//! Cumulus. It will not work outside a Cumulus parachain.
+
+use cumulus_primitives::validation_function_params::{
+	ValidationFunctionParams,
+	NEW_VALIDATION_CODE, VALIDATION_FUNCTION_PARAMS, INHERENT_IDENTIFIER,
+};
+use system::ensure_none;
+use frame_support::{
+	decl_error, decl_event, decl_module, decl_storage, ensure, storage, traits::Get,
+	weights::{SimpleDispatchInfo, Weight, MINIMUM_WEIGHT},
+};
+use parachain::primitives::RelayChainBlockNumber;
+use sp_core::storage::well_known_keys;
+use sp_inherents::{ProvideInherent, InherentData, InherentIdentifier};
+use sp_version::RuntimeVersion;
+
+pub type ValidationFunction = Vec<u8>;
+type System<T> = system::Module<T>;
+
+/// The pallet's configuration trait.
+pub trait Trait: system::Trait {
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+
+	/// Get the chain's current version.
+	type Version: Get<RuntimeVersion>;
+}
+
+// This pallet's storage items.
+decl_storage! {
+	trait Store for Module<T: Trait> as ParachainUpgrade {
+		// we need to store the new validation function for the span between
+		// setting it and applying it.
+		PendingValidationFunction get(fn new_validation_function): (RelayChainBlockNumber, ValidationFunction);
+
+		// we keep the validation function parameters here because we have to use
+		// storage as a transport layer between the inherents and the module
+		VFPs get(fn vfps): ValidationFunctionParams;
+
+		/// Were the VFPs updated this block?
+		DidUpdateVFPs: bool;
+	}
+}
+
+// The pallet's dispatchable functions.
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		// Initializing events
+		// this is needed only if you are using events in your pallet
+		fn deposit_event() = default;
+
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			if let Ok((apply_block, validation_function)) = PendingValidationFunction::try_get() {
+				let relay_chain_height = Self::validation_function_params().relay_chain_height;
```

Comment:
If there's a simple solution to this, I don't know what it is. 

The most straightforward solution I can think of right now is to add a new hook `on_inherents_complete` to `decl_module`; it would run after inherents are complete but before any non-inherents are processed. I'm pretty sure I can figure out how to do that, but it would be a substantial change to Substrate, and would mean that this PR can definitely not be released until after Substrate 2.0. Still, we could then just get rid of `on_initialize` and use `on_inherents_complete` instead.

Does that make sense, or is there a better idea that I'm not seeing?

Reply:
Actually, Substrate is weird about this and inherents can be interleaved with regular transactions as I understand it. Maybe this data could instead go in a `PreRuntime` digest (see BABE module), although it would need to go in the header then, which could be wasteful for light clients down the road.

Reply:
Would it be both possible and desirable to force all inherents to be executed before any regular transactions? I'm trying to weigh the technical benefits of my approach vs yours.

Also, I'm going to look into the weight of the `ValidationFunctionParameters`; my instinct is that it comes to less than 128 bytes, which feels negligible, but having a measurement allows us to measure precisely what impact it will have for light clients.

Reply:
In discussion, @bkchr asserted that all inherents are already executed before any transactions, which simplifies things a bit.

Reply:
For what it's worth, `ValidationFunctionParams` occupies 102 bits, assuming an 8-bit discriminant for an `Option`, not counting alignment. I don't have the knowledge to say whether that's heavy compared to the rest of the `PreRuntime` digests.

Reply:
> all inherents are already executed before any transactions, which simplifies things a bit.

This is definitely what our block producing code does, but is it enforced by the runtime somewhere?

If we have this property then we can just make this inherent mandatory, although it will be weird not to have it available in `on_initialize`. I can imagine other modules wanting hooks into `OnValidationParamsIncluded` as a replacement.

Reply:
https://github.com/paritytech/cumulus/pull/70/commits/3eb1618d2cb63b2aa186fa2694da382d3f678a3c should resolve this problem.

Reply:
I have left a comment on that snippet; I don't really understand the approach.